GVRf/Framework/framework/src/main/jni/objects/components/render_data.cpp

```diff
@@ -84,45 +89,41 @@ bool compareRenderDataByShader(RenderData* i, RenderData* j) {
 
     return i->material(0)->shader_type() < j->material(0)->shader_type();
 }
-bool compareRenderDataByOrderShaderDistance(RenderData* i,
-            RenderData* j) {
+
+bool compareRenderDataByOrderShaderDistance(RenderData *i, RenderData *j) {
     //1. rendering order needs to be sorted first to guarantee specified correct order
     if (i->rendering_order() == j->rendering_order()) {
 
         if (i->material(0)->shader_type() == j->material(0)->shader_type()) {
 
-                int no_passes1 = i->pass_count();
-                int no_passes2 = j->pass_count();
-
-                if(no_passes1 == no_passes2){
-
-                   for(int pass=0; pass < no_passes1; pass++){
-
-                        if (i->material(pass) == j->material(pass)) {
+            // if it is a transparent object, sort by camera distance from back to front
+            if (i->rendering_order() >= RenderData::Transparent
+                && i->rendering_order() < RenderData::Overlay) {
+                return i->camera_distance() > j->camera_distance();
+            }
 
-                            if(i->cull_face(pass) == j->cull_face(pass)){
+            int no_passes1 = i->pass_count();
+            int no_passes2 = j->pass_count();
 
-                                if (i->getHashCode().compare(j->getHashCode()) == 0) {
+            if (no_passes1 == no_passes2) {
 
-                                      // if it is a transparent object, sort by camera distance from back to front
-                                    if (i->rendering_order() >= RenderData::Transparent
-                                           && i->rendering_order() < RenderData::Overlay) {
-                                        return i->camera_distance() > j->camera_distance();
-                                    }
-                                        // otherwise sort from front to back
-                                    return i->camera_distance() < j->camera_distance();
-                                }
-                                return i->getHashCode() < j->getHashCode();
-                            }
-                            return i->cull_face(pass) < j->cull_face(pass);
+                //@todo what about the other passes
```

Comment:
If unclear, I removed the loop over the passes since it wasn't actually looping and could be misleading.